url11fb0973cf_raichuman9_Swimmer526_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/org/jbox2d/collision/TimeOfImpact.java

```diff
@@ -0,0 +1,549 @@
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+package org.jbox2d.collision;
+
+import org.jbox2d.collision.Distance.DistanceProxy;
+import org.jbox2d.collision.Distance.SimplexCache;
+import org.jbox2d.common.Mat22;
+import org.jbox2d.common.MathUtils;
+import org.jbox2d.common.Settings;
+import org.jbox2d.common.Sweep;
+import org.jbox2d.common.Transform;
+import org.jbox2d.common.Vec2;
+import org.jbox2d.pooling.IWorldPool;
+
+/**
+ * Class used for computing the time of impact. This class should not be
+ * constructed usually, just retrieve from the {@link SingletonPool#getTOI()}.
+ * 
+ * @author daniel
+ */
+public class TimeOfImpact {
+	public static final int MAX_ITERATIONS = 1000;
+	
+	public static int toiCalls = 0;
+	public static int toiIters = 0;
+	public static int toiMaxIters = 0;
+	public static int toiRootIters = 0;
+	public static int toiMaxRootIters = 0;
+	
+	/**
+	 * Input parameters for TOI
+	 * @author Daniel Murphy
+	 */
+	public static class TOIInput {
+		public final DistanceProxy proxyA = new DistanceProxy();
+		public final DistanceProxy proxyB = new DistanceProxy();
+		public final Sweep sweepA = new Sweep();
+		public final Sweep sweepB = new Sweep();
+		/**
+		 * defines sweep interval [0, tMax]
+		 */
+		public float tMax;
+	}
+	
+	public static enum TOIOutputState {
+		UNKNOWN, FAILED, OVERLAPPED, TOUCHING, SEPARATED
+	}
+	
+	/**
+	 * Output parameters for TimeOfImpact
+	 * @author daniel
+	 */
+	public static class TOIOutput {
+		public TOIOutputState state;
+		public float t;
+	}
+
+	
+	// djm pooling
+	private final SimplexCache cache = new SimplexCache();
+	private final DistanceInput distanceInput = new DistanceInput();
+	private final Transform xfA = new Transform();
+	private final Transform xfB = new Transform();
+	private final DistanceOutput distanceOutput = new DistanceOutput();
+	private final SeparationFunction fcn = new SeparationFunction();
+	private final int[] indexes = new int[2];
+	private final Sweep sweepA = new Sweep();
+	private final Sweep sweepB = new Sweep();
+	
+	
+	private final IWorldPool pool;
+	
+	public TimeOfImpact(IWorldPool argPool){
+		pool = argPool;
+	}
+	/**
+	 * Compute the upper bound on time before two shapes penetrate. Time is represented as
+	 * a fraction between [0,tMax]. This uses a swept separating axis and may miss some
+	 * intermediate,
+	 * non-tunneling collision. If you change the time interval, you should call this
+	 * function
+	 * again.
+	 * Note: use Distance to compute the contact point and normal at the time of impact.
+	 * 
+	 * @param output
+	 * @param input
+	 */
+	public final void timeOfImpact(TOIOutput output, TOIInput input) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 92 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.